CRM/Utils/Hook.php

```diff
@@ -152,35 +152,35 @@ public function invoke(
     &$arg1, &$arg2, &$arg3, &$arg4, &$arg5, &$arg6,
     $fnSuffix
   ) {
-    // Per https://github.com/civicrm/civicrm-core/pull/13551 we have had ongoing significant problems where hooks from modules are
-    // invoked during upgrade but not those from extensions. The issues are both that an incorrect module list & settings are cached and that
-    // some hooks REALLY need to run during upgrade - the loss of triggers during upgrade causes significant loss of data
-    // whereas loss of logTable hooks means that log tables are created for tables specifically excluded - e.g due to large
-    // quantities of low-importance data or the table not having an id field, which could cause a fatal error.
-    // Instead of not calling any hooks we only call those we know to be frequently important - if a particular extension wanted
-    // to avoid this they could do an early return on CRM_Core_Config::singleton()->isUpgradeMode
-    // Futther discussion is happening at https://lab.civicrm.org/dev/core/issues/1460
-    $upgradeFriendlyHooks = ['civicrm_alterSettingsFolders', 'civicrm_alterSettingsMetaData', 'civicrm_triggerInfo', 'civicrm_alterLogTables', 'civicrm_container', 'civicrm_permission', 'civicrm_managed', 'civicrm_config'];
-    if (CRM_Core_Config::singleton()->isUpgradeMode() && !in_array($fnSuffix, $upgradeFriendlyHooks)) {
-      return;
-    }
-    if (is_array($names) && !defined('CIVICRM_FORCE_LEGACY_HOOK') && \Civi\Core\Container::isContainerBooted()) {
-      $event = \Civi\Core\Event\GenericHookEvent::createOrdered(
-        $names,
-        array(&$arg1, &$arg2, &$arg3, &$arg4, &$arg5, &$arg6)
-      );
-      \Civi::dispatcher()->dispatch('hook_' . $fnSuffix, $event);
-      return $event->getReturnValues();
-    }
-    else {
-      // We need to ensure tht we will still run known bootstrap related hooks even if the container is not booted.
-      $prebootContainerHooks = array_merge($upgradeFriendlyHooks, ['civicrm_entityTypes', 'civicrm_config']);
-      if (!\Civi\Core\Container::isContainerBooted() && !in_array($fnSuffix, $prebootContainerHooks)) {
+    if (!\Civi\Core\Container::isContainerBooted()) {
+      $prebootHooks = ['civicrm_container', 'civicrm_entityTypes'];
```

Comment:
@totten I think we want `civicrm_alterSettingsFolders` and `civicrm_alterSettingsMetaData` as well given we want to boot with the settings as they should be modified right?

Reply:
Good question. I was a little bit surprised that this simplified as much as it did, but it checked out on each angle I could find:

1. In my testing (of the previous code) `h_c_alterSettingsFolders` and `h_c_alterSettingsMetaData` did not behave like pre-boot hooks.  Specifically, I set a breakpoint on the declaration of `$prebootContainerHooks = ...`, cleared out the `civicrm_cache` and `files/civicrm/templates_c`, and opened `civicrm/upgrade?reset=1`. The debugger showed that `h_c_container` was preboot, but that was it.

   (I also repeated the procedure on `civicrm/admin?reset=1`. That showed `h_c_entityTypes` and `h_c_container` both being preboot. `h_c_themes` popped up as a red-herring. It's not preboot per se; it's just an old `invoke(1,...)` notation; fix in #17124.)

2. In my mind/expectations, the `h_c_alterSettingsFoo` shouldn't be fired pre-boot. To resolve some chicken/egg issues while refactoring circa 4.7, I had pulled out a handful of boot-critical/early-stage-bootstrap settings into `SettingsManager::getSystemDefaults()`. This meant that the *rest* of the settings system could be deferred until late-stage bootstrap -- and then profit from a fuller set of services (extensions/hooks).

3. Thinking in terms of this PR using two-phases during upgrade, my intuition is that:
    * (a) During the planning/closed phase, I don't see how it matters if you have the settings from extensions -- even if you loaded them, they wouldn't be used, because all the other hooks are turned off. The settings are inert.
    * (b) During the finish/open phase, when it runs the system-flush, it is important to fire these hooks -- so that caches hydrate properly and so that settings are available for use within other hooks.

Reply:
So we historically had a lot of significant breakage on upgrade around settings once we migrated from modules (which always ran hooks on upgrade) to extensions. However as long as when anything more a very limited subset of settings are loaded the hooks are called & definitely before any setting metadata is cached it should be ok